examples/rotation_bin/rotation_bin/imp.rs

```diff
@@ -0,0 +1,203 @@
+                    gtk::Widget::static_type(),
+                    glib::ParamFlags::READWRITE | glib::ParamFlags::EXPLICIT_NOTIFY,
+                ),
+                glib::ParamSpecEnum::new(
+                    "rotation",
+                    "Rotation",
+                    "The Rotation of the widget",
+                    Rotation::static_type(),
+                    Rotation::default() as i32,
+                    glib::ParamFlags::READWRITE | glib::ParamFlags::EXPLICIT_NOTIFY,
+                ),
+            ]
+        });
+
+        PROPERTIES.as_ref()
+    }
+
+    fn property(&self, _obj: &Self::Type, _id: usize, pspec: &glib::ParamSpec) -> glib::Value {
+        match pspec.name() {
+            "child" => self.child.borrow().to_value(),
+            "rotation" => self.rotation.get().to_value(),
+            _ => unimplemented!(),
+        }
+    }
+
+    fn set_property(
+        &self,
+        obj: &Self::Type,
+        _id: usize,
+        value: &glib::Value,
+        pspec: &glib::ParamSpec,
+    ) {
+        match pspec.name() {
+            "child" => {
+                self.set_child(obj, value.get().ok().as_ref());
+            }
+            "rotation" => {
+                self.set_rotation(obj, value.get().unwrap());
+            }
+            _ => unimplemented!(),
+        }
+    }
+
+    fn dispose(&self, _obj: &Self::Type) {
+        if let Some(child) = self.child.borrow_mut().take() {
+            child.unparent();
+        }
+    }
+}
+
+impl WidgetImpl for RotationBin {
+    fn request_mode(&self, _widget: &Self::Type) -> SizeRequestMode {
+        if self.child.borrow().is_none() {
+            return SizeRequestMode::ConstantSize;
+        };
+
+        match self.rotation.get() {
+            Rotation::Normal | Rotation::Deg180 => SizeRequestMode::HeightForWidth,
+            Rotation::Deg90 | Rotation::Deg270 => SizeRequestMode::WidthForHeight,
+        }
+    }
+
+    fn compute_expand(&self, _widget: &Self::Type, hexpand: &mut bool, vexpand: &mut bool) {
+        let child = match self.child.borrow().as_ref() {
+            None => {
+                *hexpand = false;
+                *vexpand = false;
+                return;
+            }
+            Some(c) => c.clone(),
+        };
+
+        match self.rotation.get() {
+            Rotation::Normal | Rotation::Deg180 => {
+                *hexpand = child.compute_expand(gtk::Orientation::Vertical);
+                *vexpand = child.compute_expand(gtk::Orientation::Horizontal);
+            }
+            Rotation::Deg90 | Rotation::Deg270 => {
+                *hexpand = child.compute_expand(gtk::Orientation::Horizontal);
+                *vexpand = child.compute_expand(gtk::Orientation::Vertical);
+            }
+        };
+    }
+
+    fn measure(
+        &self,
+        _widget: &Self::Type,
+        orientation: gtk::Orientation,
+        for_size: i32,
+    ) -> (i32, i32, i32, i32) {
+        if self.child.borrow().is_none() {
```

Comment:
In all these functions, why don't you do

```rust
let child = match &*self.child.borrow() {
    None => return ...,
    Some(child) => child,
};
```

And then don't do the `unwrap` later?

Reply:
Hmm, doesn't look as pretty. The borrow makes the compiler complain about dropping it.

Reply:
How does it fail exactly and with which code?

Reply:
So say on the following code:

```
        let child = match &*self.child.borrow() {
            Some(child) => child,
            None => return (0, 0, -1, -1),
        };
 ```
 
 You get this  error
 
 ```
 error[E0716]: temporary value dropped while borrowed
   --> examples/rotation_bin/rotation_bin/imp.rs:133:29
    |
133 |         let child = match &*self.child.borrow() {
    |                             ^^^^^^^^^^^^^^^^^^^ creates a temporary which is freed while still in use
...
136 |         };
    |          - temporary value is freed at the end of this statement
...
154 |                 let (min, nat, _, _) = child.measure(orientation, for_size);
    |                                        ----- borrow later used here
    |
    = note: consider using a `let` binding to create a longer lived value
```

Which gets easily fixed with this, but then its a bit of clutter and I tend to miss what's going on inside the match or why it happens.
```
        let child = self.child.borrow();
        let child = match child.as_ref() {}
```

Anyhow the code is like this so you can take a look, should add comments on it regardless.

Reply:
Yeah read again the code I wrote above :)

```rust
let child_borrow = self.child.borrow();
let child = match &*child_borrow {
    None => return ...,
    Some(child) => child,
};
```

Reply:
You need to keep the `Ref` alive

Reply:
Your version is OK too of course, same thing

Reply:
oh indeed, tunnel vision. Anyhow, its fine either way